core/src/bls_sigverify/bls_sigverifier.rs

```diff
@@ -0,0 +1,1675 @@
+        });
+        verify_time.stop();
+        self.stats
+            .verify_elapsed_us
+            .fetch_add(verify_time.as_us(), Ordering::Relaxed);
+
+        votes_result?;
+        certs_result?;
+
+        // Send to metrics service for metrics aggregation
+        if self
+            .consensus_metrics_sender
+            .send((Instant::now(), consensus_metrics_to_send))
+            .is_err()
+        {
+            warn!("could not send consensus metrics, receive side of channel is closed");
+        }
+
+        self.stats.maybe_report_stats();
+
+        Ok(())
+    }
+
+    fn verify_and_send_votes(
+        &self,
+        votes_to_verify: Vec<VoteToVerify>,
+    ) -> Result<(), BLSSigVerifyServiceError> {
+        let verified_votes = self.verify_votes(votes_to_verify);
+        self.stats
+            .total_valid_packets
+            .fetch_add(verified_votes.len() as u64, Ordering::Relaxed);
+
+        let mut verified_votes_by_pubkey: AHashMap<Pubkey, AHashSet<Slot>> =
+            AHashMap::with_capacity(AVERAGE_VOTES_OR_CERTS_PER_BATCH);
+        self.stats
+            .received_votes
+            .fetch_add(verified_votes.len() as u64, Ordering::Relaxed);
+        for vote in verified_votes {
+            if vote.vote_message.vote.is_notarization_or_finalization()
+                || vote.vote_message.vote.is_notarize_fallback()
+            {
+                let slot = vote.vote_message.vote.slot();
+                verified_votes_by_pubkey
+                    .entry(vote.pubkey)
+                    .or_default()
+                    .insert(slot);
+            }
+
+            // Send the BLS vote messaage to certificate pool
+            match self
+                .message_sender
+                .try_send(ConsensusMessage::Vote(vote.vote_message))
+            {
+                Ok(()) => {
+                    self.stats.sent.fetch_add(1, Ordering::Relaxed);
+                }
+                Err(TrySendError::Full(_)) => {
+                    self.stats.sent_failed.fetch_add(1, Ordering::Relaxed);
+                }
+                Err(e @ TrySendError::Disconnected(_)) => {
+                    return Err(e.into());
+                }
+            }
+        }
+
+        // Send votes
+        for (pubkey, slots) in verified_votes_by_pubkey {
+            match self
+                .verified_votes_sender
+                .try_send((pubkey, slots.into_iter().collect()))
+            {
+                Ok(()) => {
+                    self.stats
+                        .verified_votes_sent
+                        .fetch_add(1, Ordering::Relaxed);
+                }
+                Err(e) => {
+                    trace!("Failed to send verified vote: {e}");
+                    self.stats
+                        .verified_votes_sent_failed
+                        .fetch_add(1, Ordering::Relaxed);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn verify_optimistically(
```

Comment:
might be nice to spell out what optimistically means here (assuming everything will verify and thus saving time by batch verifying)

Reply:
Done.